### ISSUE TITLE
macOS: Official CPython packages give "code signature invalid"

### ISSUE THREAD

**Issue by An-dz** [Nuitka/Nuitka#786]
* MacOS Catalina (10.15)
* Nuitka v0.6.8.4
* Python v3.8.3
* Nuitka installed from pip
* Python installed from official package on python.org

For some reason no Nuitka compiled package works on my MacOS. I've just tested with a `print("Hello World")` compiled with `python -m nuitka --standalone test.py` and the result is always this:

```
dyld: Library not loaded: @executable_path/Python
  Referenced from: /test.dist/./main
  Reason: no suitable image found.  Did find:
	/test.dist/./Python: code signature invalid for '/test.dist/./Python'

	/test.dist/./Python: stat() failed with errno=1
	/test.dist/Python: code signature invalid for '/test.dist/Python'

	/test.dist/Python: stat() failed with errno=1
zsh: abort      ./test
```

I found issue #672 that is similar but the proposed solution of using a venv doesn't change anything for me.

The output for `otool -L`:

```
/bin/python:
	/Library/Frameworks/Python.framework/Versions/3.8/Python (compatibility version 3.8.0, current version 3.8.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1197.1.1)
```

**Comment by kayhayen** [comment]
I am also not convinced, a venv could change anything about these binaries.

However, why would `stat` fail? Can you show a listing of the dist folder created? If that's the normal `stat`, that may be it. And also, that is where a `venv` may come in, by making a copy under your user maybe.

**Comment by An-dz** [comment]
Here are the files generated on the Hello World code.

There's no difference with or without venv, and the virtual env didn't look different from other platforms.
```
-rw-r--r--  1  3,3M  Python
-rw-r--r--  1   89K  _asyncio.so
-rw-r--r--  1   34K  _bisect.so
-rw-r--r--  1   62K  _blake2.so
-rw-r--r--  1   42K  _bz2.so
-rw-r--r--  1  161K  _codecs_cn.so
-rw-r--r--  1  169K  _codecs_hk.so
-rw-r--r--  1   45K  _codecs_iso2022.so
-rw-r--r--  1  268K  _codecs_jp.so
-rw-r--r--  1  149K  _codecs_kr.so
-rw-r--r--  1  127K  _codecs_tw.so
-rw-r--r--  1   29K  _contextvars.so
-rw-r--r--  1   28K  _crypt.so
-rw-r--r--  1   53K  _csv.so
-rw-r--r--  1  158K  _ctypes.so
-rw-r--r--  1  148K  _curses.so
-rw-r--r--  1   39K  _curses_panel.so
-rw-r--r--  1  125K  _datetime.so
-rw-r--r--  1   37K  _dbm.so
-rw-r--r--  1  414K  _decimal.so
-rw-r--r--  1   97K  _elementtree.so
-rw-r--r--  1   51K  _hashlib.so
-rw-r--r--  1   39K  _heapq.so
-rw-r--r--  1   56K  _json.so
-rw-r--r--  1   42K  _lsprof.so
-rw-r--r--  1  252K  _lzma.so
-rw-r--r--  1   65K  _multibytecodec.so
-rw-r--r--  1   37K  _multiprocessing.so
-rw-r--r--  1   29K  _opcode.so
-rw-r--r--  1  149K  _pickle.so
-rw-r--r--  1   30K  _posixshmem.so
-rw-r--r--  1   36K  _posixsubprocess.so
-rw-r--r--  1   37K  _queue.so
-rw-r--r--  1   36K  _random.so
-rw-r--r--  1   31K  _scproxy.so
-rw-r--r--  1  103K  _sha3.so
-rw-r--r--  1   48K  _sha512.so
-rw-r--r--  1  103K  _socket.so
-rw-r--r--  1  1,1M  _sqlite3.so
-rw-r--r--  1  179K  _ssl.so
-rw-r--r--  1   29K  _statistics.so
-rw-r--r--  1   69K  _struct.so
-rw-r--r--  1   85K  _tkinter.so
-rw-r--r--  1   28K  _uuid.so
-rw-r--r--  1   78K  array.so
-rw-r--r--  1   60K  audioop.so
-rw-r--r--  1   52K  binascii.so
-rw-r--r--  1   39K  fcntl.so
-rw-r--r--  1   36K  grp.so
-rw-r--r--  1  2,7M  libcrypto.1.1.dylib
-rw-r--r--  1  423K  libncursesw.5.dylib
-rw-r--r--  1   44K  libpanelw.5.dylib
-rw-r--r--  1  543K  libssl.1.1.dylib
-rw-r--r--  1  1,4M  libtcl8.6.dylib
-rw-r--r--  1  1,3M  libtk8.6.dylib
-rw-r--r--  1   77K  math.so
-rw-r--r--  1   43K  mmap.so
-rw-r--r--  1  261K  pyexpat.so
-rw-r--r--  1   54K  readline.so
-rw-r--r--  1   49K  select.so
-rw-r--r--  1   35K  termios.so
-rwxr-xr-x  1  7,5M  test
-rw-r--r--  1  1,1M  unicodedata.so
-rw-r--r--  1   58K  zlib.so
```

**Comment by kayhayen** [comment]
Actually I had wanted to see the ownership, permissions look good. Maybe Python has a strange ownership?

**Comment by An-dz** [comment]
Oh, sorry for removing that then, thought it was irrelevant as all of them are owned by my user and group. I've update the answer above.

Group `staff` is the normal "user group" on MacOS.

**Comment by kayhayen** [comment]
It seems I need access to a macOS machine, that I don't have. We could change Github workflow to go with the official Python package as well, then this should be reproducible.

**Comment by kayhayen** [comment]
Ok, so I changed the Github workflow, to use standard Python installation, which is probably closer to what we recommend people to do, and low and behold, this happens:


```

+Traceback (most recent call last):
+  File "/Users/runner/work/Nuitka/Nuitka/tests/standalone/CtypesUsing.dist/CtypesUsing.py", line 20, in <module>
+  File "/Users/runner/work/Nuitka/Nuitka/tests/standalone/CtypesUsing.dist/ctypes/__init__.py", line 7, in <module>
+ImportError: dlopen(_ctypes.so, 2): image not found
```

So for me at least, the Python library, with 3.8 on Github loaded, but something very similar happens.

The file is of course there, according to the listing the standalone tests give:

> -rw-r--r--   1 runner  staff   145764 Jul 20 12:41 _ctypes.so

**Comment by kayhayen** [comment]
That was a regression from something else, it seems to work now, so I don't know what to make out of it.

Can you check this https://github.com/Nuitka/Nuitka/runs/890932329?check_suite_focus=true and tell me what can be the difference?

Yours,
Kay

**Comment by kayhayen** [comment]
Also, please try the factory branch. Possibly, I have fixed it while making string operations safe.

https://nuitka.net/doc/factory.html

**Comment by An-dz** [comment]
Unfortunately no change with factory. I'll check later what's different from the scons output and check how and where that Python file comes from.

**Comment by An-dz** [comment]
Well, one difference from my build versus yours on GitHub actions is that I'm building a **standalone** and you are building a **module**.

When compiling the Hello World code with `--module` it works. Any other way it doesn't.

But I did find something, this Python file that is copied into the dist directory is the Python executable itself and not a library file like on Windows. There's a symbolic link named `lib/libpython3.8.dylib` that points to it.

As a test I changed `freezer/Standalone.py` with the following and it worked:
```diff
 if python_version >= 300:
                 filename = filename.decode("utf-8")
 
+            if filename == '/Library/Frameworks/Python.framework/Versions/3.8/Python':
+                filename = '/Library/Frameworks/Python.framework/Versions/3.8/lib/libpython3.8.dylib'
 
             # print("adding", filename)
             result.add(filename)
```

Craziest of all is that the `libpython3.8.dylib` file on the dist directory is not a symbolic link, but a working library file.

**Comment by kayhayen** [comment]
You got it, very nearly. Can you check where the `Python` filename came from in the first place. If you search for `otool`, you will find in the freezer code, and I expect this to be about something like that tool being different.

Just to summarize, in your installation, the "libpython" filename is a symlink to "Python". That is an executable, which is fine in POSIX, every shared library is allowed to be executable, and some will e.g. output their version or license, if you do that, I have seen this before, and as such it was a mistake while Nuitka created extensions that were having the "x" bit. So using that is not forbidden.

However, when loading the shared library, maybe macOS hates it if it's not done with the correct name, so renaming a shared library will give that error. That is an assumption that would explain things here.

I am going to make file listings of these folders on Github, to see if there is another situation for their 3.8, as they are also using as 3.8, and either this is them having a different installation, which is possible of course, or their `otool` handling symlinks differently. Maybe there is an option that controls it? It is almost as if `otool` of yours did kind of `realpath` or we do it, and that breaks things.
I am also going to add otool output for the Python binary, to the Github workflow, to see what it says.

**Comment by kayhayen** [comment]
So here we are:

```
+ otool -L /Users/runner/hostedtoolcache/Python/3.8.3/x64/bin/python
/Users/runner/hostedtoolcache/Python/3.8.3/x64/bin/python:
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1575.235.0)
	/Users/runner/hostedtoolcache/Python/3.8.3/x64/lib/libpython3.8.dylib (compatibility version 3.8.0, current version 3.8.0)
	/usr/local/opt/gettext/lib/libintl.8.dylib (compatibility version 10.0.0, current version 10.7.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.250.1)
```

**Comment by kayhayen** [comment]
Notice how this is very different from yours, and probably is a result of how they configured it.

So, what we do on Windows for the Python DLL is to use this kind of code:

```
def getRunningPythonDLLPath():
    import ctypes.wintypes

    MAX_PATH = 4096
    buf = ctypes.create_unicode_buffer(MAX_PATH)

    GetModuleFileName = ctypes.windll.kernel32.GetModuleFileNameW
    GetModuleFileName.argtypes = (
        ctypes.wintypes.HANDLE,
        ctypes.wintypes.LPWSTR,
        ctypes.wintypes.DWORD,
    )
    GetModuleFileName.restype = ctypes.wintypes.DWORD

    # We trust ctypes internals here, pylint: disable=protected-access
    res = GetModuleFileName(ctypes.pythonapi._handle, buf, MAX_PATH)
    if res == 0:
        # Windows only code, pylint: disable=I0021,undefined-variable
        raise WindowsError(
            ctypes.GetLastError(), ctypes.FormatError(ctypes.GetLastError())
        )

    dll_path = os.path.normcase(buf.value)
    assert os.path.exists(dll_path), dll_path

    return dll_path
```

That looks up the Python DLL from the running program. I think there is something that allows to ask which module an entry is coming from, and I have used that before, but I cannot remember right now.

**Comment by kayhayen** [comment]
Can you verify, if you have this in your `otool -l` output?

```
Load command 13
          cmd LC_LOAD_DYLIB
      cmdsize 96
         name /Users/runner/hostedtoolcache/Python/3.8.3/x64/lib/libpython3.8.dylib (offset 24)
   time stamp 2 Thu Jan  1 00:00:02 1970
      current version 3.8.0
```

Or if you have the `Python` binary instead. Because if you don't have this, it would something otool does, although I doubt it considers the filenames at all, probably not even for existance.

**Comment by kayhayen** [comment]
Ok, theirs is not even a symlink, it's just there.

**Comment by An-dz** [comment]
> Can you verify, if you have this in your `otool -l` output?
It does not. Just lists the `Python` binary.

> Ok, theirs is not even a symlink, it's just there.
So the GitHub actions version is not exactly the same as the one from <python.org>.

> Craziest of all is that the `libpython3.8.dylib` file on the dist directory is not a symbolic link, but a working library file.
The file is the executable itself, cheksum matches, but here it's renamed. So you are probably correct that on MacOS it has to be a `dylib` file to work as a dll and the symlink is there exactly for that.

I'm not sure how this can be fixed without a direct check like my hack – which does not work correctly as `otool -L <my_executable>` shows it's linked after the `Python` executable under the original installation.

**Comment by kayhayen** [comment]
We can definitely include a workaround for something like this. I increasingly regret to not have a system, where I can test this.

I could abuse Github if I knew how to install via command line. But I can also guess make a patch that handles the situation, which is what I will attempt now. Making this uber-perfect is not really needed indeed.

Yours,
Kay

**Comment by kayhayen** [comment]
So on factory, there is your patch, just with checking for existance and version independent.

**Comment by kayhayen** [comment]
Please try this out and let me know if it works unchanged, or I have some kind of error in that change: http://nuitka.net/doc/factory.html

**Comment by An-dz** [comment]
`lib_filename` is undefined, it won't work. But anyway my patch (and yours by extension) does not work because the link is done on the executable in the standard installation (under `/Library`) and is not changed to the library file so it's not a real standalone package.

The problem is that when `fixupBinaryDLLPathsMacOS` is called the filenames are already modified and it tries to replace a link to `/Library/Frameworks/Python.framework/Versions/<version>/lib/libpython<version>.dylib` to the one in the dist directory, but there's no link to that file in the generated executable, only a link to the executable under `/Library/Frameworks/Python.framework/Versions/<version>/Python`.

I tried changing that moving the code to `copyUsedDLLs`, but to no avail, even though I managed to change the link to a relative path the error is still the same.

I'm starting to think it's impossible to get this working with a standard Python installation.

**Comment by gauthierbuttez** [comment]
Hi,

I have this exact same issue:
```
Margauxs-MacBook:~ margaux$ /Users/margaux/Documents/MyApp_0002/MyApp.dist/MyApp ; exit;
dyld: Library not loaded: @executable_path/Python
  Referenced from: /Users/margaux/Documents/MyApp_0002/MyApp.dist/MyApp
  Reason: no suitable image found.  Did find:
	/Users/margaux/Documents/MyApp_0002/MyApp.dist/Python: code signature invalid for '/Users/margaux/Documents/MyApp_0002/MyApp.dist/Python'

	/Users/margaux/Documents/MyApp_0002/MyApp.dist/Python: stat() failed with errno=1
	/Users/margaux/Documents/MyApp_0002/MyApp.dist/Python: code signature invalid for '/Users/margaux/Documents/MyApp_0002/MyApp.dist/Python'

	/Users/margaux/Documents/MyApp_0002/MyApp.dist/Python: stat() failed with errno=1
Abort trap: 6
logout
Saving session...
...copying shared history...
...saving history...truncating history files...
...completed.
Deleting expired sessions...3 completed.

[Process completed]
```

My environment is :
macOS Mojave 10.14.16
Python 3.8
but when I type nuitka --version:
```
0.6.8.4
Python: 2.7.10 (default, Feb 22 2019, 21:55:15) 
Executable: /usr/bin/python
OS: Darwin
Arch: x86_64
```
You noticed the Python version displayed by Nuitka => 2.7. But when I type "python --version" I get 3.8.5.
So I recompiled Nuitka as I downloaded the source:
`sudo python setup.py install`
But it still displays "Python: 2.7.10 (default, Feb 22 2019, 21:55:15) '"

Here are some more info:
```
otool -L /usr/lib/python
python:
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1570.13.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.250.1)
```

```
otool -L /Library/Frameworks/Python.framework/Versions/3.8/Python
/Library/Frameworks/Python.framework/Versions/3.8/Python:
	/Library/Frameworks/Python.framework/Versions/3.8/Python (compatibility version 3.8.0, current version 3.8.0)
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 855.17.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1197.1.1)
```


I wanted to try your factory version but the links are dead:
https://github.com/Nuitka/Nuitka/archive/
http://git.nuitka.net/Nuitka.git

Let me know if you need any extra information to handle this kind of scenario.

**Comment by kayhayen** [comment]
I wonder where you got those links from @gauthierbuttez

**Comment by kayhayen** [comment]
Without a machine to test on, I am very unlikely to be able to fix this, help is needed.

**Comment by kayhayen** [comment]
Fixed, see #789 it seems the code signature must be removed or else it will not be possible to load the DLL.

**Comment by kayhayen** [comment]
The code is now part of the latest pre-release.

**Comment by An-dz** [comment]
Will try it out on the weekend or earlier. Thanks a lot for fix.

**Comment by kayhayen** [comment]
Released as part of 0.6.12
